transport/transporttest/choice.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transporttest
+
+import (
+	"go.uber.org/yarpc/transport"
+
+	"github.com/golang/mock/gomock"
+)
+
+// NewChoiceMatcher returns a pair of matchers corresponding to the arguments
+// of GetHandler(Context, Request), for verifying that GetHandler is called with
+// parameters that satisfy given constraints. Passing options like
+// WithService("foo") adds constraints to the matcher.
+func NewChoiceMatcher(choiceOptions ...choiceOption) (gomock.Matcher, gomock.Matcher) {
+	cm := choiceMatcher{
+		constraints: make([]choiceConstraint, 0, 5),
+	}
+	for _, choiceOption := range choiceOptions {
+		choiceOption(&cm)
+	}
+	return gomock.Any(), &cm
+}
+
+// WithCaller returns an option for NewChoiceMatcher that adds a constraint that
+// the request must have the given caller.
+func WithCaller(caller string) func(*choiceMatcher) {
+	return func(cm *choiceMatcher) {
+		cm.constraints = append(cm.constraints, func(r *transport.Request) bool {
+			return caller == r.Caller
+		})
+	}
```

Comment:
IMO it would be simpler to just have a `*ChoiceMatcher` object with methods on it that mutate it rather than taking these names up globally.

```
NewChoiceMatcher().WithCaller("foo").WithService("bar")
```